File: python/tanque.py
from abc import ABC
from estructura import Estructura


class Tanque(Estructura, ABC):

    def __init__(self, capacidad_total):
        Estructura.__init__(self)
        self._capacidad_total = capacidad_total
        self._capacidad_disponible = capacidad_total

    def capacidad_total(self):
        return self._capacidad_total

    def capacidad_disponible(self):
        return self._capacidad_disponible

    def cargar(self, volumen_a_cargar):
        volumen_cargado = min([volumen_a_cargar, self.capacidad_disponible()])
        self._capacidad_disponible -= volumen_cargado
        return volumen_cargado

    def extraer(self, volumen_a_extraer):
        volumen_extraido = min([volumen_a_extraer, self.capacidad_total() - self.capacidad_disponible()])
        self._capacidad_disponible += volumen_extraido
        return volumen_extraido

File: java/src/simoil/Tanque.java
package simoil;


public class Tanque  {
    private String nombre;
    private double capacidadTotal;
    private double capacidadDisponible;

    public Tanque(String nombre, double capacidadTotal) {
        this.nombre = nombre;
        if (capacidadTotal <= 0)
            throw new RuntimeException("La capacidad de un tanque debe ser positiva.");
        this.capacidadTotal = capacidadTotal;
        this.capacidadDisponible = capacidadTotal;
    }

    public double capacidadTotal() {
        return capacidadTotal;
    }

    public double capacidadDisponible() {
        return capacidadDisponible;
    }

    public double volumenCargado() {
        return capacidadTotal - capacidadDisponible;
    }

    public double cargar(double volumenACargar) {
        double volumenCargado = Math.min(volumenACargar, capacidadDisponible);
        capacidadDisponible -= volumenCargado;
        return volumenCargado;
    }

    public double descargar(double volumenADescargar) {
        double volumenDescargado = Math.min(volumenADescargar, volumenCargado());
        capacidadDisponible += volumenDescargado;
        return volumenDescargado;
    }

    public double descargarTodo() {
        double volumenDescargado = volumenCargado();
        capacidadDisponible = capacidadTotal;
        return volumenDescargado;
    }

    public String nombre(){
        return this.nombre;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Tanque tanque = (Tanque) o;

        return nombre.equals(tanque.nombre);
    }

}
